File: packages/core/src/lexer.js
/**
 * @license
 * Copyright (c) 2025 LEAN Format Team and contributors
 * Licensed under MIT License
 */

export const TokenType = {
    INDENT: 'INDENT',
    DEDENT: 'DEDENT',
    NEWLINE: 'NEWLINE',
    IDENTIFIER: 'IDENTIFIER',
    STRING: 'STRING',
    NUMBER: 'NUMBER',
    BOOLEAN: 'BOOLEAN',
    NULL: 'NULL',
    COLON: 'COLON',
    HYPHEN: 'HYPHEN',
    COMMA: 'COMMA',
    LPAREN: 'LPAREN',
    RPAREN: 'RPAREN',
    EOF: 'EOF'
};

export class Token {
    constructor(type, value, line, column) {
        this.type = type;
        this.value = value;
        this.line = line;
        this.column = column;
    }
}

export class Lexer {
    constructor(input) {
        this.input = input.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
        this.pos = 0;
        this.line = 1;
        this.column = 1;
        this.indentStack = [0];
        this.tokens = [];
        this.indentationHandled = false;
        this.isStartOfLine = true;
        this.indentChar = null;
    }

    tokenize() {
        while (this.pos < this.input.length) {
            const char = this.peek();

            // Handle newlines
            if (char === '\n') {
                this.addToken(TokenType.NEWLINE, '\n');
                this.advance();
                this.line++;
                this.column = 1;
                this.indentationHandled = false;
                this.isStartOfLine = true;
                continue;
            }

            // Handle indentation at start of line
            if (this.column === 1 && !this.indentationHandled) {
                this.handleIndentation();
                this.indentationHandled = true;
                if (this.pos >= this.input.length) break;
                // Re-peek after indentation handling as pos might have changed
                continue;
            }

            // Skip whitespace (except newlines)
            if (char === ' ' || char === '\t') {
                this.advance();
                continue;
            }

            // Handle comments
            // Only treat # as comment if it's the first non-whitespace character (handled by handleIndentation check)
            // OR if we are at the start of the line (column 1)
            // But wait, handleIndentation consumes indentation.
            // So if we are here, and char is #, it IS the first non-whitespace char.
            // UNLESS we are in the middle of a line.
            // We need to know if we are at the start of the line (after indentation).
            // We can check if `this.column` equals `indentation + 1`? No.
            // We can check if we just handled indentation?
            // Or simply: The Lexer loop handles tokens.
            // If we see `#`, is it a comment?
            // If inline comments are NOT supported, then `#` is only a comment if it's at the start of the line.
            // But `handleIndentation` skips whitespace.
            // So if `   # comment`, `handleIndentation` runs. `pos` at `#`.
            // Then `char` is `#`.
            // If `key: value # not comment`.
            // `key`, `:`, `value`, ` `. `char` is `#`.
            // Here `column` is NOT at start.
            // So we should treat `#` as a character (part of identifier/string) if it's not at start.

            if (char === '#') {
                // Check if we are at the start of the line (ignoring indentation)
                // We can check if the previous tokens on this line were only INDENT?
                // Or easier: check if `this.column` matches the expected indentation + 1?
                // But we don't track "expected indentation".

                // Actually, `handleIndentation` sets `this.indentationHandled = true`.
                // And `column` > 1.
                // But if we are at the start of the line, we just handled indentation (or column is 1).
                // So if we are at the start, we should skip comment.
                // But if we parsed other tokens on this line, we should NOT skip.

                // We can track `tokensOnLine` count? Reset on NEWLINE.
                // If `tokensOnLine === 0` (ignoring INDENT), then `#` is comment.

                // But wait, `INDENT` is a token.
                // So if `tokensOnLine === 0` or `tokensOnLine === 1` (INDENT).

                // Let's add `tokensOnLine` to Lexer.

                // For now, let's just implement the logic:
                // If char is #, and we are NOT at start of line (meaning we have seen other tokens), treat as symbol?
                // But `handleIndentation` runs first.
                // If `   #`, `handleIndentation` runs. `pos` at `#`.
                // We need to know if we are "at start".

                // Let's use a flag `isStartOfLine`.
                // Set to true on NEWLINE.
                // Set to false when we emit a token (other than INDENT/DEDENT).

                // But `INDENT` is emitted.
                // So if we emit `INDENT`, `isStartOfLine` remains true?
                // Yes, because `#` after indent is still a comment.

                // So:
                // 1. Add `isStartOfLine` flag. Init true.
                // 2. On NEWLINE, set true.
                // 3. On `addToken`: if type is NOT INDENT/DEDENT/NEWLINE/EOF, set false.
                // 4. If `char === '#'`, check `isStartOfLine`.
                //    If true, skip comment.
                //    If false, treat as punctuation/symbol (or part of identifier).

                // Wait, `#` is not in `TokenType`.
                // So if it's not a comment, what is it?
                // It falls through to `readIdentifier` or `readString`?
                // `readIdentifier` regex: `/[a-zA-Z0-9_$.-]/`. `#` is not there.
                // So it throws "Unexpected character".

                // So we need to support `#` in identifiers?
                //            // Handle comments
                if (char === '#') {
                    this.skipComment();
                    continue;
                }    // But we need to ensure `#` is handled.
                // We can add `#` to `isIdentifierPart`?
            }

            // Handle punctuation
            if (char === ':') {
                this.addToken(TokenType.COLON, ':');
                this.advance();
                continue;
            }
            if (char === '-') {
                // Check if it's a negative number or a hyphen
                if (this.isNextDigit()) {
                    this.readNumber();
                } else {
                    this.addToken(TokenType.HYPHEN, '-');
                    this.advance();
                }
                continue;
            }
            if (char === ',') {
                this.addToken(TokenType.COMMA, ',');
                this.advance();
                continue;
            }
            if (char === '(') {
                this.addToken(TokenType.LPAREN, '(');
                this.advance();
                continue;
            }
            if (char === ')') {
                this.addToken(TokenType.RPAREN, ')');
                this.advance();
                continue;
            }

            // Handle strings
            if (char === '"') {
                this.readString();
                continue;
            }

            // Handle numbers
            if (this.isDigit(char)) {
                this.readNumber();
                continue;
            }

            // Handle identifiers and keywords
            if (this.isIdentifierStart(char)) {
                this.readIdentifier();
                continue;
            }

            throw new Error(`Unexpected character '${char}' at line ${this.line}, column ${this.column}`);
        }

        // Dedent to zero at EOF
        while (this.indentStack.length > 1) {
            this.indentStack.pop();
            this.addToken(TokenType.DEDENT, '');
        }

        this.addToken(TokenType.EOF, '');
        return this.tokens;
    }

    handleIndentation() {
        let indentLevel = 0;
        // indentChar is initialized in constructor

        // Calculate indentation
        let currentPos = this.pos;
        while (currentPos < this.input.length) {
            const char = this.input[currentPos];
            if (char === ' ' || char === '\t') {
                if (this.indentChar === null) {
                    this.indentChar = char;
                } else if (char !== this.indentChar) {
                    throw new Error(`Mixed indentation (spaces and tabs) at line ${this.line}`);
                }

                if (char === ' ') {
                    indentLevel++;
                } else { // tab
                    indentLevel += 4; // Treat tab as 4 spaces for calculation
                }
            } else {
                break;
            }
            currentPos++;
        }

        // If line is empty or comment-only, ignore indentation
        if (currentPos >= this.input.length || this.input[currentPos] === '\n' || this.input[currentPos] === '#') {
            const indentLength = currentPos - this.pos;
            this.pos = currentPos;
            this.column += indentLength;
            // If it's a comment, skip it, but don't reset column yet as we might be on a new line next
            return;
        }

        const currentIndent = this.indentStack[this.indentStack.length - 1];

        if (indentLevel > currentIndent) {
            this.indentStack.push(indentLevel);
            this.addToken(TokenType.INDENT, indentLevel);
        } else if (indentLevel < currentIndent) {
            while (this.indentStack.length > 1 && this.indentStack[this.indentStack.length - 1] > indentLevel) {
                this.indentStack.pop();
                this.addToken(TokenType.DEDENT, '');
            }
            // The old parser logic for mixed indentation is now handled above
            // if (whitespace[i] !== this.indentChar) {
            //     // throw new Error('Mixed indentation (spaces and tabs)', this.currentLine);
            //     // Wait, this logic was in old parser. Lexer handles it differently.
            // }
        }
        // Lexer logic:
        // We need to check consistency in handleIndentation?
        // But handleIndentation just counts spaces/tabs.
        // It doesn't enforce "only spaces" or "only tabs" across the file?
        // The old parser did: `if (whitespace[i] !== this.indentChar)`.
        // My Lexer doesn't track `indentChar` globally yet.

        // I should add `indentChar` tracking to Lexer to support this check.

        // But for now, let's just fix the error message in `handleIndentation` if inconsistent.
        // Wait, `handleIndentation` throws `Inconsistent indentation`.
        // This is when indentation level doesn't match stack.

        // The test expects "Mixed indentation (spaces and tabs)".
        // This implies mixing spaces and tabs in the SAME line or across file.
        // My Lexer treats tab as 4 spaces.
        // It doesn't check for mixed chars.

        // I should implement mixed indentation check.
        // Add `indentChar` to Lexer. Init null.
        // In `handleIndentation`, check first char.
        // If `indentChar` is null, set it.
        // If `indentChar` is set, check if current indentation uses same char.

        // But `handleIndentation` iterates `currentPos`.
        // `char = this.input[currentPos]`.
        // If `char` is space or tab.

        // Let's update `handleIndentation` to check mixed indentation.

        const indentLength = currentPos - this.pos;
        this.pos = currentPos;
        this.column += indentLength;
    }

    skipComment() {
        while (this.pos < this.input.length && this.peek() !== '\n') {
            this.advance();
        }
    }

    readString() {
        let value = '';
        this.advance(); // Skip opening quote

        while (this.pos < this.input.length) {
            const char = this.peek();
            if (char === '"') {
                this.advance(); // Skip closing quote
                this.addToken(TokenType.STRING, value);
                return;
            }
            if (char === '\\') {
                this.advance();
                const escape = this.peek();
                const escapes = { 'n': '\n', 'r': '\r', 't': '\t', '\\': '\\', '"': '"' };
                value += escapes[escape] || escape;
                this.advance();
            } else {
                value += char;
                this.advance();
            }
        }
        throw new Error(`Unterminated string at line ${this.line}`);
    }

    readNumber() {
        let value = '';
        if (this.peek() === '-') {
            value += '-';
            this.advance();
        }

        while (this.pos < this.input.length && (this.isDigit(this.peek()) || this.peek() === '.' || this.peek() === 'e' || this.peek() === 'E')) {
            value += this.peek();
            this.advance();
        }

        this.addToken(TokenType.NUMBER, parseFloat(value));
    }

    readIdentifier() {
        let value = '';
        while (this.pos < this.input.length && this.isIdentifierPart(this.peek())) {
            value += this.peek();
            this.advance();
        }

        if (value === 'true') this.addToken(TokenType.BOOLEAN, true);
        else if (value === 'false') this.addToken(TokenType.BOOLEAN, false);
        else if (value === 'null') this.addToken(TokenType.NULL, null);
        else if (value.includes('.')) {
            // Handle dot notation as a single identifier for now, parser will split
            this.addToken(TokenType.IDENTIFIER, value);
        } else {
            // Treat unquoted strings as identifiers/strings depending on context
            // For now, everything else is an identifier which can be a key or an unquoted string value
            this.addToken(TokenType.IDENTIFIER, value);
        }
    }

    peek() {
        return this.input[this.pos];
    }

    advance() {
        this.pos++;
        this.column++;
    }

    addToken(type, value) {
        this.tokens.push(new Token(type, value, this.line, this.column));
        if (type !== TokenType.INDENT && type !== TokenType.DEDENT && type !== TokenType.NEWLINE && type !== TokenType.EOF) {
            this.isStartOfLine = false;
        }
    }

    isDigit(char) {
        return /[0-9]/.test(char);
    }

    isNextDigit() {
        return this.pos + 1 < this.input.length && this.isDigit(this.input[this.pos + 1]);
    }

    isIdentifierStart(char) {
        return /[a-zA-Z_$]/.test(char);
    }

    isIdentifierPart(char) {
        return char === '#' || /[-a-zA-Z0-9_$.@]/.test(char);
    }
}

File: packages/lean-parser-rs/src/lexer.rs


#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    Indent,
    Dedent,
    Newline,
    Identifier,
    String,
    Number,
    Boolean,
    Null,
    Colon,
    Hyphen,
    Comma,
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    Eof,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenValue {
    String(String),
    Number(f64),
    Boolean(bool),
    Null,
    None,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub token_type: TokenType,
    pub value: TokenValue,
    pub line: usize,
    pub column: usize,
}

pub struct Lexer {
    input: Vec<char>,
    pos: usize,
    pub line: usize,
    pub column: usize,
    indent_stack: Vec<usize>,
    pub tokens: Vec<Token>,
    indentation_handled: bool,
    is_start_of_line: bool,
    indent_char: Option<char>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        // Replace \r\n and \r with \n
        let normalized = input.replace("\r\n", "\n").replace('\r', "\n");
        Self {
            input: normalized.chars().collect(),
            pos: 0,
            line: 1,
            column: 1,
            indent_stack: vec![0],
            tokens: Vec::new(),
            indentation_handled: false,
            is_start_of_line: true,
            indent_char: None,
        }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, String> {
        while self.pos < self.input.len() {
            let ch = self.peek().unwrap();

            if ch == '\n' {
                self.add_token(TokenType::Newline, TokenValue::String("\n".to_string()));
                self.advance();
                self.line += 1;
                self.column = 1;
                self.indentation_handled = false;
                self.is_start_of_line = true;
                continue;
            }

            if self.column == 1 && !self.indentation_handled {
                self.handle_indentation()?;
                self.indentation_handled = true;
                if self.pos >= self.input.len() {
                    break;
                }
                continue;
            }

            if ch == ' ' || ch == '\t' {
                self.advance();
                continue;
            }

            if ch == '#' {
                self.skip_comment();
                continue;
            }

            if ch == ':' {
                self.add_token(TokenType::Colon, TokenValue::String(":".to_string()));
                self.advance();
                continue;
            }

            if ch == '-' {
                if self.is_next_digit() {
                    self.read_number()?;
                } else {
                    self.add_token(TokenType::Hyphen, TokenValue::String("-".to_string()));
                    self.advance();
                }
                continue;
            }

            if ch == ',' {
                self.add_token(TokenType::Comma, TokenValue::String(",".to_string()));
                self.advance();
                continue;
            }

            if ch == '(' {
                self.add_token(TokenType::LParen, TokenValue::String("(".to_string()));
                self.advance();
                continue;
            }

            if ch == ')' {
                self.add_token(TokenType::RParen, TokenValue::String(")".to_string()));
                self.advance();
                continue;
            }

            if ch == '{' {
                self.add_token(TokenType::LBrace, TokenValue::String("{".to_string()));
                self.advance();
                continue;
            }

            if ch == '}' {
                self.add_token(TokenType::RBrace, TokenValue::String("}".to_string()));
                self.advance();
                continue;
            }

            if ch == '[' {
                self.add_token(TokenType::LBracket, TokenValue::String("[".to_string()));
                self.advance();
                continue;
            }

            if ch == ']' {
                self.add_token(TokenType::RBracket, TokenValue::String("]".to_string()));
                self.advance();
                continue;
            }

            if ch == '"' {
                self.read_string()?;
                continue;
            }

            if Self::is_digit(ch) {
                self.read_number()?;
                continue;
            }

            if Self::is_identifier_start(ch) {
                self.read_identifier();
                continue;
            }

            return Err(format!("Unexpected character '{}' at line {}, column {}", ch, self.line, self.column));
        }

        while self.indent_stack.len() > 1 {
            self.indent_stack.pop();
            self.add_token(TokenType::Dedent, TokenValue::String("".to_string()));
        }

        self.add_token(TokenType::Eof, TokenValue::String("".to_string()));
        Ok(self.tokens.clone())
    }

    fn handle_indentation(&mut self) -> Result<(), String> {
        let mut indent_level = 0;
        let mut current_pos = self.pos;

        while current_pos < self.input.len() {
            let ch = self.input[current_pos];
            if ch == ' ' || ch == '\t' {
                if self.indent_char.is_none() {
                    self.indent_char = Some(ch);
                } else if Some(ch) != self.indent_char {
                    return Err(format!("Mixed indentation (spaces and tabs) at line {}", self.line));
                }

                if ch == ' ' {
                    indent_level += 1;
                } else {
                    indent_level += 4;
                }
            } else {
                break;
            }
            current_pos += 1;
        }

        if current_pos >= self.input.len() || self.input[current_pos] == '\n' || self.input[current_pos] == '#' {
            let indent_length = current_pos - self.pos;
            self.pos = current_pos;
            self.column += indent_length;
            return Ok(());
        }

        let current_indent = *self.indent_stack.last().unwrap();

        if indent_level > current_indent {
            self.indent_stack.push(indent_level);
            // In TS, INDENT token value is indentLevel
            self.add_token(TokenType::Indent, TokenValue::Number(indent_level as f64));
        } else if indent_level < current_indent {
            while self.indent_stack.len() > 1 && *self.indent_stack.last().unwrap() > indent_level {
                self.indent_stack.pop();
                self.add_token(TokenType::Dedent, TokenValue::String("".to_string()));
            }
        }

        let indent_length = current_pos - self.pos;
        self.pos = current_pos;
        self.column += indent_length;
        Ok(())
    }

    fn skip_comment(&mut self) {
        while self.pos < self.input.len() && self.peek().unwrap() != '\n' {
            self.advance();
        }
    }

    fn read_string(&mut self) -> Result<(), String> {
        let mut value = String::new();
        self.advance(); // Skip opening quote

        while self.pos < self.input.len() {
            let ch = self.peek().unwrap();
            if ch == '"' {
                self.advance(); // Skip closing quote
                self.add_token(TokenType::String, TokenValue::String(value));
                return Ok(());
            }
            if ch == '\\' {
                self.advance();
                if self.pos >= self.input.len() {
                    return Err(format!("Unterminated string at line {}", self.line));
                }
                let escape = self.peek().unwrap();
                match escape {
                    'n' => value.push('\n'),
                    'r' => value.push('\r'),
                    't' => value.push('\t'),
                    '\\' => value.push('\\'),
                    '"' => value.push('"'),
                    _ => value.push(escape),
                }
                self.advance();
            } else {
                value.push(ch);
                self.advance();
            }
        }
        Err(format!("Unterminated string at line {}", self.line))
    }

    fn read_number(&mut self) -> Result<(), String> {
        let mut value = String::new();
        if self.peek().unwrap() == '-' {
            value.push('-');
            self.advance();
        }

        while self.pos < self.input.len() {
            let ch = self.peek().unwrap();
            if Self::is_digit(ch) || ch == '.' || ch == 'e' || ch == 'E' || ch == '-' || ch == '+' {
                // '-' and '+' allowed for scientific notation (e.g., e-5)
                value.push(ch);
                self.advance();
            } else {
                break;
            }
        }

        if let Ok(num) = value.parse::<f64>() {
            self.add_token(TokenType::Number, TokenValue::Number(num));
            Ok(())
        } else {
            Err(format!("Invalid number format '{}' at line {}", value, self.line))
        }
    }

    fn read_identifier(&mut self) {
        let mut value = String::new();
        while self.pos < self.input.len() {
            let ch = self.peek().unwrap();
            if Self::is_identifier_part(ch) {
                value.push(ch);
                self.advance();
            } else {
                break;
            }
        }

        match value.as_str() {
            "true" => self.add_token(TokenType::Boolean, TokenValue::Boolean(true)),
            "false" => self.add_token(TokenType::Boolean, TokenValue::Boolean(false)),
            "null" => self.add_token(TokenType::Null, TokenValue::Null),
            _ => self.add_token(TokenType::Identifier, TokenValue::String(value)),
        }
    }

    fn peek(&self) -> Option<char> {
        if self.pos < self.input.len() {
            Some(self.input[self.pos])
        } else {
            None
        }
    }

    fn advance(&mut self) {
        self.pos += 1;
        self.column += 1;
    }

    fn add_token(&mut self, token_type: TokenType, value: TokenValue) {
        self.tokens.push(Token {
            token_type: token_type.clone(),
            value,
            line: self.line,
            column: self.column,
        });
        if !matches!(token_type, TokenType::Indent | TokenType::Dedent | TokenType::Newline | TokenType::Eof) {
            self.is_start_of_line = false;
        }
    }

    fn is_digit(ch: char) -> bool {
        ch.is_ascii_digit()
    }

    fn is_next_digit(&self) -> bool {
        if self.pos + 1 < self.input.len() {
            Self::is_digit(self.input[self.pos + 1])
        } else {
            false
        }
    }

    fn is_identifier_start(ch: char) -> bool {
        ch.is_ascii_alphabetic() || ch == '_' || ch == '$'
    }

    fn is_identifier_part(ch: char) -> bool {
        ch.is_ascii_alphanumeric() || ch == '_' || ch == '$' || ch == '.' || ch == '-'
    }
}
